### ISSUE TITLE
[java] UseDiamondOperator misses a few reports where diamond could be used

### ISSUE THREAD

**Issue by aaime** [pmd/pmd#3353]
**Affects PMD Version:** 6:35

**Rule:** [UseDiamondOperator](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#usediamondoperator)

The rule is declared as:

```
  <rule ref="category/java/codestyle.xml/UseDiamondOperator" />
```

In other words, there is Java 7 specific behavior is not enabled, should be performing checks for Java 8+ type inference.

**Description:**

An automated refactoring tool called [OpenRewrite](https://docs.openrewrite.org/) found a few cases where the diamond syntax could be used, which PMD did not catch. See the example below:

**Code Sample demonstrating the issue:**

```java
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.function.Consumer;

public class DiamondIssues {

    public static void main(final String[] args) {
        // <String> is not necessary. IntelliJ does not catch this one either
        doStuff(new ArrayList<String>());

        // specifying <Date> is also not necessary. IntelliJ does not catch this one either
        final List<Timestamp> timestamps = Arrays.asList(new Timestamp(new Date().getTime()));
        doOtherStuff(new ArrayList<Date>(timestamps));
        
        // <Integer> is not required. IntelliJ correctly reports this one as un-necessary
        addConsumer(Integer.class, new MyConsumer<Integer>());
    }

    private static void doStuff(final Object object) {
        System.out.println(object);
    }

    private static void doOtherStuff(final List<? extends Date> dates) {
        System.out.println(dates);
    }

    private static <T> void addConsumer(final Class<? extends T> type, final Consumer<T> consumer) {
        // do something useful here
    }

    private static class MyConsumer<T extends Number> implements Consumer<T> {

        @Override
        public void accept(final T t) {
            System.out.println(t);
        }
    }
}

```

**Expected outcome:**

PMD should report a violation at line 14, 18 and 21, but doesn't. This is a false-negative.
Maybe worth mentioning, the first two are not reported by IntelliJ either, while the third is correctly spotted.

**Running PMD through:** *[Maven]*
